### DISCUSSION TITLE
Custom modification S3 upload adapter

### DISCUSSION THREAD

**Discussion by angelhodar** [directus/directus#5951] [Feature Requests]
Hey, I just wanted to know if there is a way to customize the way a file is uploaded from the admin app. For example, it would be extremely helpful in my use case if I could upload a specific folder to a specific bucket name and the rest of the files to another (or just change the type of adapter in a dropdown below the dropzone). I have seen that @rijkvanzanten and @benhaynes were talking about this [here](https://github.com/directus/directus/discussions/3076), but I dont know if it is still in the roadmap.

Also, is it possible to change the name of the uploaded file in S3? I would like to keep the original name of the file, not the generated uuid.  I have used strapi and they allow to easily customize S3 behaivour by creating a custom adapter. 

Im curious about why did you guys decided to include the adapters inside the main code instead of making it independent modules that the user can import as needed. Anyway thanks in advance!!!

**Comment by rijkvanzanten** [comment]
Heya!

> For example, it would be extremely helpful in my use case if I could upload a specific folder to a specific bucket name

You can use the "root" environment variable to set the specific folder to upload to in the configured bucket (https://docs.directus.io/reference/environment-variables/#file-storage)

> [...] and the rest of the files to another

The API already supports this (you can provide the `storage` field to tell it where to upload the file to. The app doesn't have a selection for the storage adapters yet, but this will be coming in the future.

> I would like to keep the original name of the file

This sounds like a good new feature request to me 👍🏻 

> Im curious about why did you guys decided to include the adapters inside the main code instead of making it independent modules that the user can import as needed

They are actually built as individual modules that can be exposed as an extension type in the future 👍🏻 In previous releases, we learned that there's really only a handful adapters that were actually used (local / s3 / gcs / azure), so it makes most sense to us to focus on having those core ones as polished as possible, instead of trying to rely on people to hack in their own solutions for everything non-local.

**Reply by angelhodar** [reply]
Hey! Thank you for your fast answer

> You can use the "root" environment variable to set the specific folder to upload to in the configured bucket (https://docs.directus.io/reference/environment-variables/#file-storage)

I was refering to select the folder in a per file basis in a bucket itself

> The API already supports this (you can provide the storage field to tell it where to upload the file to. The app doesn't have a selection for the storage adapters yet, but this will be coming in the future.

Perfect! Just a dropdown when uploading files would do the trick.

> This sounds like a good new feature request to me 👍🏻

In my use case is because I need to consume the bucket uploads from another app and it needs to keep the file names.

> They are actually built as individual modules that can be exposed as an extension type in the future 👍🏻 In previous releases, we learned that there's really only a handful adapters that were actually used (local / s3 / gcs / azure), so it makes most sense to us to focus on having those core ones as polished as possible, instead of trying to rely on people to hack in their own solutions for everything non-local.

My question was more related to why you decided to have them by default in the directus package instead of writing in the documentation the npm packages available so the user can install them depending on their needs. For example someone may not need S3 support and they have to include the whole aws-sdk (well if you upload it to the v3 you only require the S3 modules, but thats not the point of my question hahaha).

**Reply by rijkvanzanten** [reply]
> so the user can install them depending on their needs

The packages are all listed under `optionalDependencies`, so by using `--no-optional` in npm, they're skipped 👍🏻

**Reply by angelhodar** [reply]
Then its perfect! Will you guys allow to customize adapter behaviour in the future with the extensions folder? Like Strapi does including functions in module.exports that override adapter behaivour.

**Reply by rijkvanzanten** [reply]
Most likely!

**Reply by angelhodar** [reply]
I have just seen that directus storage adapters are not listed in the optionalDependencies, they are just below the dependencies in the api package. Am I wrong?

**Reply by rijkvanzanten** [reply]
Ah that's a mistake then, those should be optional 👍🏻 Wanna open a PR to rectify that? (We'd probably want to keep the `local` one included, as that's the default)

**Reply by angelhodar** [reply]
Sure! I will do it later. I suppose I also have to update the documentation and probably add some checks to the env parsing because I dont know if it assumes that al adapters are available. If you want I can also try to implement UI updates to select the adapter on a per file basis (probably in another PR).

**Reply by rijkvanzanten** [reply]
Yeah, we'll have to make this bit require the correct libraries dynamically, instead of relying on the top level import:

https://github.com/directus/directus/blob/521dfd3f8b3292b037cf229dc69194e2d625e0f5/api/src/storage.ts#L59-L70

**Comment by jaads** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
